sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_do_while.py

```diff
@@ -54,4 +55,28 @@ def pipeline_with_do_while_node():
         _from_component_func=True,
     )
     do_while_node.set_limits(max_iteration_count=max_iteration_count)
+
+    def _infer_and_update_body_input_from_mapping():
+        # pylint: disable=protected-access
+        for source_output, body_input in mapping.items():
+            # handle case that mapping key is a NodeOutput
+            output_name = source_output.port_name if isinstance(source_output, NodeOutput) else source_output
+            # if loop body output type is not specified, skip as we have no place to infer
+            if body.outputs[output_name].type is None:
+                continue
+            # if input type is specified, no need to infer and skip
+            if body_input.type is not None:
+                continue
+            inferred_type = body.outputs[output_name].type
+            # update node input
+            body_input._meta._is_inferred_optional = True
+            body_input.type = inferred_type
+            # update node corresponding component input
+            input_name = body_input._meta.name
+            body.component.inputs[input_name]._is_inferred_optional = True
```

Comment:
When do we need to infer optional?

Reply:
For dynamic input, the property optional will be None, and in following logic, it will be treated as required, which may lead to validation fail for the loop body (required input not provided); and as this optional is inferred by SDK code, so update inferred optional` to True rather than original optional field.